### PR TITLE
#17515 Avoid clash between field vars and graphql content fields

### PR DESCRIPTION
**Pull request by dsilvam**
Since any GraphQL Type inherits a list of fields from the GraphQL `Contentlet` Interface, there was a possibility of name-clashing if you named a Content Type field the same as one of those inherited fields. 
With this PR those inherited field names are included in the list of taken variable names to avoid the clashing. 

Tests were included for each reserved field name to verify the resulting variable is different than the reserved name.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldFactoryImpl.java [hunk 1]
```diff
@@ -232,8 +234,13 @@ private Field dbSaveUpdate(final Field throwAwayField) throws DotDataException {
       }
 
       // normalize our velocityvar
+      final List<String> takenFieldVars = fieldsAlreadyAdded.stream().map(Field::variable).collect(
+              Collectors.toList());
+      // let's add GraphQL reserved field names to the taken fields vars list
+      takenFieldVars.addAll(InterfaceType.RESERVED_GRAPHQL_FIELD_NAMES);
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldFactoryImpl.java]: maybe we should move this logic to its own Util, something like:
```
class GrahpQLUtil {
    public List<String> getReservedWord(){
         return InterfaceType.RESERVED_GRAPHQL_FIELD_NAMES;
   }
}
```

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldFactory.java [hunk 2]
```diff
@@ -51,7 +51,7 @@ default FieldFactory instance() {
     List<Field> selectByContentTypeInDb(String id) throws DotDataException;
 
 
-    String suggestVelocityVar(String tryVar, List<Field> takenFields) throws DotDataException;
+    String suggestVelocityVar(String tryVar, List<String> takenFields) throws DotDataException;
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldFactory.java]: doc, why are you changing this method signature?, if the List<String> is a list with the Field's id (or some other field attribute) I think is better keep the List of Fied and into the method get the field's id

**Review by freddyucv — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldFactory.java (hunk 2 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/FieldFactory.java]: The only thing we used from Field was its name

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 3]
```diff
@@ -1035,13 +1036,64 @@ public void testSaveFieldVariable_KeyWithSpecialChars_ShouldSucceed(final String
...
+                    .name(fieldName)
+                    .contentTypeId(type.id())
+                    .indexed(false)
+                    .listed(false)
+                    .fixed(true)
+                    .build();
+            field1 = fieldAPI.save(field1, user);
+
+            Assert.assertNotNull(field1);
+            Assert.assertTrue(UtilMethods.isSet(field1.variable()));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.assertTrue'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 4]
```diff
@@ -1035,13 +1036,64 @@ public void testSaveFieldVariable_KeyWithSpecialChars_ShouldSucceed(final String
...
+            Field field2 = FieldBuilder.builder(TextField.class)
+                    .name(fieldName)
+                    .contentTypeId(type.id())
+                    .indexed(false)
+                    .listed(false)
+                    .fixed(true)
+                    .build();
+            field2 = fieldAPI.save(field2, user);
+
+            Assert.assertNotNull(field2);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertNotNull' due to existing static import 'org.junit.Assert.assertNotNull'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 5]
```diff
@@ -1035,13 +1036,64 @@ public void testSaveFieldVariable_KeyWithSpecialChars_ShouldSucceed(final String
...
+            Assert.assertTrue(UtilMethods.isSet(field2.variable()));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.assertTrue'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 6]
```diff
@@ -1035,13 +1036,64 @@ public void testSaveFieldVariable_KeyWithSpecialChars_ShouldSucceed(final String
...
+            Field field1 = FieldBuilder.builder(TextField.class)
+                    .name(fieldName)
+                    .contentTypeId(type.id())
+                    .indexed(false)
+                    .listed(false)
+                    .fixed(true)
+                    .build();
+            field1 = fieldAPI.save(field1, user);
+
+            Assert.assertNotNull(field1);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertNotNull' due to existing static import 'org.junit.Assert.assertNotNull'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldFactoryImpl.java (hunk 1 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/FieldFactoryImpl.java]: Done

**Review by nollymar — APPROVED** [review]

**Review by freddyucv — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/graphql/util/GraphQLUtil.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.dotcms.graphql.util;
+
+import com.dotcms.graphql.InterfaceType;
+import java.util.Set;
+
+public class GraphQLUtil {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/graphql/util/GraphQLUtil.java]: Codacy Issue found: [All methods are static.  Consider using a utility class instead. Alternatively, you could add a private constructor or make the class abstract to silence this warning.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 8]
```diff
@@ -23,6 +23,8 @@
 import com.dotcms.contenttype.model.type.SimpleContentType;
 import com.dotcms.datagen.ContentTypeDataGen;
 import com.dotcms.datagen.TestDataUtils;
+import com.dotcms.graphql.InterfaceType;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.graphql.InterfaceType'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 9]
```diff
@@ -1035,13 +1037,64 @@ public void testSaveFieldVariable_KeyWithSpecialChars_ShouldSucceed(final String
...
+            Field field1 = FieldBuilder.builder(TextField.class)
+                    .name(fieldName)
+                    .contentTypeId(type.id())
+                    .indexed(false)
+                    .listed(false)
+                    .fixed(true)
+                    .build();
+            field1 = fieldAPI.save(field1, user);
+
+            Assert.assertNotNull(field1);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertNotNull' due to existing static import 'org.junit.Assert.assertNotNull'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 10]
```diff
@@ -1035,13 +1037,64 @@ public void testSaveFieldVariable_KeyWithSpecialChars_ShouldSucceed(final String
...
+            Assert.assertTrue(UtilMethods.isSet(field1.variable()));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.assertTrue'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 11]
```diff
@@ -1035,13 +1037,64 @@ public void testSaveFieldVariable_KeyWithSpecialChars_ShouldSucceed(final String
...
+            Field field2 = FieldBuilder.builder(TextField.class)
+                    .name(fieldName)
+                    .contentTypeId(type.id())
+                    .indexed(false)
+                    .listed(false)
+                    .fixed(true)
+                    .build();
+            field2 = fieldAPI.save(field2, user);
+
+            Assert.assertNotNull(field2);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertNotNull' due to existing static import 'org.junit.Assert.assertNotNull'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 12]
```diff
@@ -1035,13 +1037,64 @@ public void testSaveFieldVariable_KeyWithSpecialChars_ShouldSucceed(final String
...
+            Assert.assertTrue(UtilMethods.isSet(field2.variable()));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.assertTrue'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 8 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.graphql.InterfaceType'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 10 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.assertTrue'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 11 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertNotNull' due to existing static import 'org.junit.Assert.assertNotNull'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 9 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertNotNull' due to existing static import 'org.junit.Assert.assertNotNull'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/main/java/com/dotcms/graphql/util/GraphQLUtil.java (hunk 7 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/graphql/util/GraphQLUtil.java]: Codacy Issue found: [All methods are static.  Consider using a utility class instead. Alternatively, you could add a private constructor or make the class abstract to silence this warning.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 12 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.assertTrue'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4926848)

**Pull request merged**